tools/clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -15690,6 +15720,61 @@ void TryAddShaderAttrFromTargetProfile(Sema &S, FunctionDecl *FD,
   return;
 }
 
+// in the non-library case, this function will be run only once,
+// but in the library case, this function will be run for each
+// viable top-level function declaration by
+// ValidateNoRecursionInTranslationUnit.
+//  (viable as in, is exported)
+clang::FunctionDecl *ValidateNoRecursion(clang::Sema *self,
+                                         clang::FunctionDecl *FD) {
+  // Validate that there is no recursion reachable by this function declaration
+  // NOTE: the information gathered here could be used to bypass code generation
+  // on functions that are unreachable (as an early form of dead code
+  // elimination).
+  if (FD) {
+    hlsl::CallGraphWithRecurseGuard CG;
+    CG.BuildForEntry(FD);
+    FunctionDecl *pResult = CG.CheckRecursion(FD);
+    if (pResult) {
```

Comment:
nit: If we're returning nullptr anyway by default, there's no need to check the result of pResult for null before returning it.